### PR TITLE
ops: revise tpetra matrix-vector product

### PR DESCRIPTION
**Pull request by mzuzek**
refs #446

### CODE REVIEW

File: include/pressio/ops/tpetra/ops_level2.hpp [hunk 1]
```diff
@@ -127,10 +127,19 @@ _product_tpetra_mv_sharedmem_vec_kokkos(const alpha_t & alpha,
...
+     ::pressio::Traits<A_type>::rank == 2
+  && ::pressio::Traits<x_type>::rank == 1
+  && ::pressio::Traits<y_type>::rank == 1
+  // TPL/container specific
   && ::pressio::is_multi_vector_tpetra<A_type>::value
   && ::pressio::is_vector_tpetra<y_type>::value
   && ::pressio::is_dense_vector_teuchos<x_type>::value
+  // scalar compatibility
+  && ::pressio::all_have_traits_and_same_scalar<A_type, x_type, y_type>::value
+  && std::is_convertible<alpha_t, typename ::pressio::Traits<A_type>::scalar_type>::value
```
Comment (fnrizzi) [inline include/pressio/ops/tpetra/ops_level2.hpp:140]: where is the check for beta?

File: include/pressio/ops/tpetra/ops_level2.hpp [hunk 2]
```diff
@@ -152,10 +161,19 @@ product(::pressio::nontranspose /*unused*/,
...
+     ::pressio::Traits<A_type>::rank == 2
+  && ::pressio::Traits<x_type>::rank == 1
+  && ::pressio::Traits<y_type>::rank == 1
+  // TPL/container specific
   && ::pressio::is_multi_vector_tpetra<A_type>::value
   && ::pressio::is_vector_tpetra<y_type>::value
   && ::pressio::is_vector_kokkos<x_type>::value
+  // scalar compatibility
+  && ::pressio::all_have_traits_and_same_scalar<A_type, x_type, y_type>::value
+  && std::is_convertible<alpha_t, typename ::pressio::Traits<A_type>::scalar_type>::value
```
Comment (fnrizzi) [inline include/pressio/ops/tpetra/ops_level2.hpp:174]: where is the check for beta?

File: include/pressio/ops/tpetra/ops_level2.hpp [hunk 3]
```diff
@@ -180,10 +198,19 @@ product(::pressio::nontranspose /*unused*/,
...
+     ::pressio::Traits<A_type>::rank == 2
+  && ::pressio::Traits<x_type>::rank == 1
+  && ::pressio::Traits<y_type>::rank == 1
+  // TPL/container specific
   && ::pressio::is_multi_vector_tpetra<A_type>::value
   && ::pressio::is_vector_tpetra<y_type>::value
   && ::pressio::is_vector_eigen<x_type>::value
+  // scalar compatibility
+  && ::pressio::all_have_traits_and_same_scalar<A_type, x_type, y_type>::value
+  && std::is_convertible<alpha_t, typename ::pressio::Traits<A_type>::scalar_type>::value
```
Comment (fnrizzi) [inline include/pressio/ops/tpetra/ops_level2.hpp:211]: where is the check for beta?

File: include/pressio/ops/tpetra/ops_level2.hpp [hunk 4]
```diff
@@ -207,10 +234,19 @@ product(::pressio::nontranspose /*unused*/,
...
+  && ::pressio::Traits<x_type>::rank == 1
+  && ::pressio::Traits<y_type>::rank == 1
+  // TPL/container specific
   && ::pressio::is_multi_vector_tpetra<A_type>::value
   && ::pressio::is_vector_tpetra<y_type>::value
-  && ::pressio::is_vector_eigen<x_type>::value,
+  && ::pressio::is_vector_eigen<x_type>::value
+// scalar compatibility
+  && ::pressio::all_have_traits_and_same_scalar<A_type, x_type, y_type>::value
+  && std::is_convertible<alpha_t, typename ::pressio::Traits<A_type>::scalar_type>::value
```
Comment (fnrizzi) [inline include/pressio/ops/tpetra/ops_level2.hpp:247]: where is the check for beta?

File: include/pressio/ops/tpetra/ops_level2.hpp [hunk 5]
```diff
@@ -226,6 +262,60 @@ product(::pressio::nontranspose mode,
...
+     ::pressio::Traits<A_type>::rank == 2
+  && ::pressio::Traits<x_type>::rank == 1
+  && ::pressio::Traits<y_type>::rank == 1
+  // TPL/container specific
+  && ::pressio::is_multi_vector_tpetra<A_type>::value
+  && ::pressio::is_vector_tpetra<y_type>::value
+  && ::pressio::is_expression_acting_on_eigen<x_type>::value
+  // scalar compatibility
+  && ::pressio::all_have_traits_and_same_scalar<A_type, x_type, y_type>::value
+  && std::is_convertible<alpha_t, typename ::pressio::Traits<A_type>::scalar_type>::value
```
Comment (fnrizzi) [inline include/pressio/ops/tpetra/ops_level2.hpp:284]: where is the check for beta?

File: include/pressio/ops/tpetra/ops_level2.hpp [hunk 6]
```diff
@@ -235,10 +325,20 @@ product(::pressio::nontranspose mode,
...
+  && ::pressio::Traits<y_type>::rank == 1
+  // TPL/container specific
   && ::pressio::is_multi_vector_tpetra<A_type>::value
   && ::pressio::is_vector_tpetra<x_type>::value
-  && ::pressio::is_vector_eigen<y_type>::value
+  && (::pressio::is_vector_eigen<y_type>::value
+   || ::pressio::is_expression_acting_on_eigen<y_type>::value)
+  // scalar compatibility
+  && ::pressio::all_have_traits_and_same_scalar<A_type, x_type, y_type>::value
+  && std::is_convertible<alpha_t, typename ::pressio::Traits<A_type>::scalar_type>::value
```
Comment (fnrizzi) [inline include/pressio/ops/tpetra/ops_level2.hpp:339]: where is the check for beta?

File: include/pressio/ops/tpetra/ops_level2.hpp [hunk 7]
```diff
@@ -277,10 +378,19 @@ product(::pressio::transpose /*unused*/,
...
+     ::pressio::Traits<A_type>::rank == 2
+  && ::pressio::Traits<x_type>::rank == 1
+  && ::pressio::Traits<y_type>::rank == 1
+  // TPL/container specific
   && ::pressio::is_multi_vector_tpetra<A_type>::value
   && ::pressio::is_vector_tpetra<x_type>::value
   && ::pressio::is_vector_kokkos<y_type>::value
+  // scalar compatibility
+  && ::pressio::all_have_traits_and_same_scalar<A_type, x_type, y_type>::value
+  && std::is_convertible<alpha_t, typename ::pressio::Traits<A_type>::scalar_type>::value
```
Comment (fnrizzi) [inline include/pressio/ops/tpetra/ops_level2.hpp:391]: where is the check for beta?

File: include/pressio/ops/tpetra/ops_level2.hpp (hunk 1 above)
Comment (mzuzek) [reply include/pressio/ops/tpetra/ops_level2.hpp:140]: thanks for catching this and pardon the omission... fixed on ee8991317e67d004ea8dace6fd23071e0215762c

File: include/pressio/ops/tpetra/ops_level2.hpp (hunk 4 above)
Comment (mzuzek) [reply include/pressio/ops/tpetra/ops_level2.hpp:250]: there is no beta here :wink:

**Comment by mzuzek** [comment]
@fnrizzi 

### Test coverage

Tpetra level-2 implementations of `pressio::ops::product()` are tested with `Tpetra::MultiVector` matrix following combinations of `x`/`y` vectors:

| `x` | `y` | mode |
|:---|:---|:---|
| eigen:<br>• vector<br>• `span` expression<br>• `diag` expression | tpetra | `nontranspose` |
| tpetra | eigen:<br>• vector<br>• `span` expression<br>• `diag` expression | `transpose` |
| kokkos:<br>• vector<br>• `span` expression<br>• `diag` expression | tpetra | `nontranspose` |
| tpetra | kokkos:<br>• vector<br>• `span` expression<br>• `diag` expression  | `transpose` |
| teuchos<br>(`Teuchos::SerialDenseVector`) | tpetra | `nontranspose` |

File: include/pressio/ops/tpetra/ops_level2.hpp [hunk 8]
```diff
@@ -226,6 +301,61 @@ product(::pressio::nontranspose mode,
...
+  const auto y2_h = y.getLocalViewHost(Tpetra::Access::ReadWriteStruct());
+  const auto y_h = Kokkos::subview(y2_h, Kokkos::ALL(), 0);
+
+  assert( ::pressio::ops::extent(y_h, 0) == ::pressio::ops::extent(A_h, 0) );
+  assert( ::pressio::ops::extent(x, 0) == ::pressio::ops::extent(A_h, 1) );
+
+  const auto zero = ::pressio::utils::Constants<beta_t>::zero();
+  if (beta == zero) ::pressio::ops::set_zero(y_h);
+  else ::pressio::ops::scale(y_h, beta);
+  if (alpha == zero) return;
```
Comment (fnrizzi) [inline include/pressio/ops/tpetra/ops_level2.hpp:345]: minor comment but from now on, can you please write these as: I like to have braces for readability 
```
const auto zero = ::pressio::utils::Constants<beta_t>::zero();
  if (beta == zero){ 
    ::pressio::ops::set_zero(y_h);
  }
  else{ 
    ::pressio::ops::scale(y_h, beta);
  }
  if (alpha == zero){ 
    return;
   }
```

File: include/pressio/ops/tpetra/ops_level2.hpp [hunk 9]
```diff
@@ -263,7 +404,8 @@ product(::pressio::transpose /*unused*/,
       // colI is a Teuchos::RCP<Vector<...>>
       const auto colI = A.getVector(i);
       y(i) = beta == zero ? zero : beta * y(i);
-      y(i) += alpha * colI->dot(x);
+      if (!(alpha == zero))
+        y(i) += alpha * colI->dot(x);
```
Comment (fnrizzi) [inline include/pressio/ops/tpetra/ops_level2.hpp:408]: can you please put braces here ? 
```
if () {
  y(i) += ... 
}

**Review by fnrizzi — CHANGES_REQUESTED** [review]
see comments

File: include/pressio/ops/tpetra/ops_level2.hpp (hunk 8 above)
Comment (mzuzek) [reply include/pressio/ops/tpetra/ops_level2.hpp]: ok, fixed on b4d7832b8b1f7bb4b120acedbad85454936b8f40

PS. we have #116 open for auto-formatting,
and now `clang-format` 15 can do `InsertBraces: true` and [other cool things](https://clang.llvm.org/docs/ClangFormatStyleOptions.html)

File: include/pressio/ops/tpetra/ops_level2.hpp [hunk 10]
```diff
@@ -67,6 +67,7 @@ namespace impl{
 template <class A_type, class x_type, class y_type, class alpha_t, class beta_t>
 ::pressio::mpl::enable_if_t<
   ::pressio::is_multi_vector_tpetra<A_type>::value and
+  ::pressio::is_vector_eigen<x_type>::value and
   ::pressio::is_vector_tpetra<y_type>::value
```
Comment (fnrizzi) [inline include/pressio/ops/tpetra/ops_level2.hpp:71]: this function is just an impl, and the name already conveys what it is doing, I think we can remove these two lines
```cpp
//remove this ::pressio::is_multi_vector_tpetra<A_type>::value and 
//remove this ::pressio::is_vector_eigen<x_type>::value and
::pressio::is_vector_tpetra<y_type>::value
```

**Review by fnrizzi — APPROVED** [review]

File: include/pressio/ops/tpetra/ops_level2.hpp (hunk 10 above)
Comment (fnrizzi) [reply include/pressio/ops/tpetra/ops_level2.hpp:71]: we can leave this for now , not a blocker

File: include/pressio/ops/tpetra/ops_level2.hpp [hunk 11]
```diff
@@ -96,6 +97,7 @@ _product_tpetra_mv_sharedmem_vec(const alpha_t & alpha,
 template <class A_type, class x_type, class y_type, class alpha_t, class beta_t>
 ::pressio::mpl::enable_if_t<
   ::pressio::is_multi_vector_tpetra<A_type>::value and
+  ::pressio::is_vector_kokkos<x_type>::value and
   ::pressio::is_vector_tpetra<y_type>::value
   >
 _product_tpetra_mv_sharedmem_vec_kokkos(const alpha_t & alpha,
```
Comment (fnrizzi) [inline include/pressio/ops/tpetra/ops_level2.hpp:103]: please remove `_kokkos`

File: include/pressio/ops/tpetra/ops_level2.hpp [hunk 12]
```diff
@@ -115,6 +118,46 @@ _product_tpetra_mv_sharedmem_vec_kokkos(const alpha_t & alpha,
...
 }
+
+template <class A_type, class x_type, class y_type, class alpha_t, class beta_t>
+::pressio::mpl::enable_if_t<
+  ::pressio::is_multi_vector_tpetra<A_type>::value and
+  (::pressio::is_expression_acting_on_kokkos<x_type>::value or
+   ::pressio::is_dense_vector_teuchos<x_type>::value) and
+  ::pressio::is_vector_tpetra<y_type>::value
+  >
+_product_tpetra_mv_sharedmem_vec_kokkos(const alpha_t & alpha,
```
Comment (fnrizzi) [inline include/pressio/ops/tpetra/ops_level2.hpp:129]: same comment as above

**Review by fnrizzi — CHANGES_REQUESTED** [review]
see comments

File: include/pressio/ops/tpetra/ops_level2.hpp (hunk 11 above)
Comment (mzuzek) [reply include/pressio/ops/tpetra/ops_level2.hpp]: thanks for catching this, fixed on 7247774710e3c0b08d6081210e957d81a602b3f3

**Comment by mzuzek** [comment]
@fnrizzi thank you for your review and comments!

**Pull request merged**